tools/c7n_mailer/c7n_mailer/ldap_lookup.py

```diff
@@ -0,0 +1,146 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+import redis
+import re
+import shelve
+from ldap3 import Connection
+from ldap3.core.exceptions import LDAPSocketOpenError
+
+
+class LdapLookup(object):
+
+    def __init__(self, config, logger, get_connection=True):
+        # This is to allow us to instantiate this object and
+        # run the instance methods in unit tests
+        if get_connection:
+            self.connection = self.get_connection(
+                config.get('ldap_uri'),
+                config.get('ldap_bind_user', None),
+                config.get('ldap_bind_password', None)
+            )
+        self.log          = logger
+        self.base_dn      = config.get('ldap_bind_dn')
+        self.email_key    = config.get('ldap_email_key', 'mail')
+        manager_attr      = config.get('ldap_manager_attribute', 'manager')
+        self.attributes   = ['displayName', self.email_key, manager_attr]
+        self.uid_key      = config.get('ldap_username_attribute', 'sAMAccountName')
+        self.uid_regex    = config.get('ldap_uid_regex', None)
+        self.cache_engine = config.get('cache_engine', None)
+        if self.cache_engine == 'redis':
+            redis_host = config.get('redis_host')
+            redis_port = int(config.get('redis_port', 6379))
+            self.caching = redis.StrictRedis(host=redis_host, port=redis_port, db=0)
+        elif self.cache_engine == 'localshelf':
+            self.caching = LocalShelf('/tmp/ldap.cache')
+
+    def get_connection(self, ldap_uri, ldap_bind_user, ldap_bind_password):
+        # note, if ldap_bind_user and ldap_bind_password are None
+        # an anonymous bind will be attempted.
+        try:
+            conn = Connection(
+                ldap_uri, user=ldap_bind_user, password=ldap_bind_password,
+                auto_bind=True,
+                receive_timeout=30,
+                auto_referrals=False,
+            )
+        except LDAPSocketOpenError:
+            conn = None
+        return conn
+
+    def get_email_from_metadata(self, ldap_user_metadata):
+        return ldap_user_metadata.get(self.email_key, None)
+
+    def user_metadata_has_manager(self, ldap_user_metadata):
+        return ldap_user_metadata.get('manager', None)
+
+    def get_metadata_from_dn(self, user_dn):
+        if self.cache_engine:
+            cache_result = self.caching.get(user_dn, None)
+            if cache_result:
+                return cache_result
+        ldap_filter = '(%s=*)' % self.uid_key
+        self.connection.search(user_dn, ldap_filter, attributes=self.attributes)
+        ldap_user_metadata = self.get_dict_from_ldap_object(self.connection.entries[0])
+        if self.cache_engine:
+            self.caching.set(user_dn, ldap_user_metadata)
+        return ldap_user_metadata
+
+    def get_toaddr_from_ldap_metadata(self, ldap_user_to_email_map, ldap_username, email_manager):
```

Comment:
a helper method to let you get the user email and the manager email in an array read to be set as the 'to' field in the Mimetext message. This is a convenient helper method for the SqsMessageProcessor object to use.